Created module for RHEL 7 to RHEL 8 upgrade

bangelic opened the pull request:
Created a new module for the master.adoc for Upgrading and Updating Foreman. This module is titled proc_upgrading-el7-to-el8.adoc. The content came from an existing document and was cleaned up for minimalist purposes and grammar. The content was pulled from this link: https://hackmd.io/@evgeni/B1AT256Q5

@melcorr This is the one you needed a PR for. Needed to clean up the content to fit minimalist principles. Did not see any attributes for RHEL so let me know if I missed that.

Cherry-pick into:

* [ ] Foreman 3.2
* [ ] Foreman 3.1
* For Foreman 3.0 or older, file a separate PR request

<!---
Thank you for contributing to Foreman documentation. Make sure to read README
for the documentation standards. Set cherry-pick github label to mark this
contribution for cherry picking and check which version do you need with [x].
-->

melcorr commented:
The attribute for RHEL is {RHEL}

bangelic commented:
@melcorr Updated with {RHEL} attribute.

melcorr commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+
+.Prerequisites
+* {Project} {ProjectVersionPrevious} running on CentOS 7.
+- Upgrading does not work on {RHEL}. You must have a {RHEL}-compatible LEAPP build to upgrade.
```
I do not understand the use of a dash here. 
We have a strict one sentence per line rule and have always had. Please adhere everywhere and always.

melcorr commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+- Upgrading does not work on {RHEL}. You must have a {RHEL}-compatible LEAPP build to upgrade.
+* Access to a local repository server mirroring all required repositories.
+* No outstanding updates.
+* You can upgrade a CentOS7 system before you upgrade {Project}.
```
Please follow conventions when referring to an operating system. This is not how it is written.

bangelic replied:
Some of the other writers mentioned that they have used CentOS7 when documenting. What is the proper naming convention for an operating system?

melcorr replied:
If you google and find the CentOS 7 documentation or website, you'll find how to write it.

melcorr replied:
Brian, I am sorry. I think that I came across quite short here in retrospect and it wasn't my intention. All I meant is that there are certain ways that all operating systems, projects, third party infrastructure etc can and should be referred to specifically and it's easy enough to find them. Their project or product websites are the authority on how it should be spelled rather than myself or any writer.

melcorr commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+or:
++
+----
+ # package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers can install
```
why is there text in the code block?

bangelic replied:
It appears in the code block in the document that was written to be converted to Asciidoc.

melcorr replied:
I think that @evgeni meant this as a comment

bangelic replied:
Thanks!

evgeni replied:
I meant this as "you could see errors like … or …" and didn't know how to format it better as "this is the error output"

melcorr commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+ # package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers can install
+----
+
+** Remove packages that are no longer available from any repository:
```
I'm pretty sure that an ordered list would be better if these are a sequence of steps.

evgeni replied:
this step is not required with satellite

melcorr commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+# dnf remove $(dnf repoquery --extras --exclude '*katello*' --exclude '*pulp*' --exclude '*localhost*' --exclude "*$HOSTNAME*" --exclude libmodulemd) –
+----
+
++
```
What is this for?

bangelic replied:
The order of the procedure was steps 1-12 and I had to add the + to keep the continuing list from restarting back at 1 in the middle of the steps.

melcorr commented on guides/doc-Upgrading_and_Updating/master.adoc:
```diff
@@ -56,6 +56,11 @@ include::topics/upgrading_discovery_satellite.adoc[leveloffset=+3]
 // Configuring Subnets with a Template Capsule
 include::topics/upgrading_discovery_subnet_with_a_template_capsule.adoc[leveloffset=+3]
 
+ifdef::katello,satellite[]
```
@evgeni 
would you mind advising which parts are not applicable for Satellite? in the procedure in general?

evgeni replied:
this applies to foreman on EL too, so should be `ifdef::foreman-el,katello,satellite` I think

melcorr reviewed (changes requested):
Some comments.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+[id="Upgrading-Project-from-EL7-to-EL8{context}"]
+= Upgrading {Project} from {RHEL} 7 to {RHEL} 8
+
+Use this procedure to upgrade your {Project} installation from {RHEL} 7 to {RHEL} 8.
+LEAPP is the supported tool for {RHEL} 7 to {RHEL} 8 upgrades.
```
Here you say upgrading from RHEL7 to RHEL8, later you write it needs to run on CentOS 7 and upgrades on RHEL are not possible. This will confuse all readers :)

Just to make it clear:
* Foreman and Katello installations running on CentOS 7 can be upgraded to CentOS Stream 8 (and probably any other EL8-rebuild like Rocky and Alma, even tho I didn't test that explicitly).
* Foreman and Katello installations running on RHEL7 can't be upgraded (right now).
* Satellite installations will be upgradeable from RHEL7 to RHEL8, no CentOS/EL-rebuilds.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+.Prerequisites
+* {Project} {ProjectVersionPrevious} running on CentOS 7.
+- Upgrading does not work on {RHEL}. You must have a {RHEL}-compatible LEAPP build to upgrade.
+* Access to a local repository server mirroring all required repositories.
```
No. The system needs access to the repositories we offer, *OR* a local mirror.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+# yum install leapp leapp-repository leapp-data-centos
+----
+
+. Add {Project} specific repositories to `/etc/leapp/files/leapp_upgrade_repositories.repo`:
```
please have a look at https://github.com/theforeman/theforeman.org/pull/1977, I've updated this section with the "client" repo, which should be reflected here

evgeni replied:
this step is not required for satellite

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+or:
++
+----
+ # package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers can install
```
a `#` usually indicates a command, but this is just a text block

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+
+. Reboot the system.
+. After the system reboots, a live system conducts the upgrade.
+. Reboot to fix SELinux labels.
+. Reboot into the final {RHEL} 8 system.
```
these steps happen automaticaly, the user doesn't need to do anything

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+. After the system reboots, a live system conducts the upgrade.
+. Reboot to fix SELinux labels.
+. Reboot into the final {RHEL} 8 system.
+. After the system reboots, LEAPP finishes the upgrade.
+. You can watch LEAPP finish the upgrade with:
```
this looks like two steps, but the "watch" part is more of a comment to the "leapp does finish"

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+* You can upgrade a CentOS7 system before you upgrade {Project}.
+
+.Procedure
+. Configure the https://copr.fedorainfracloud.org/coprs/g/theforeman/leapp/[@theforeman/leapp COPR Repository], which contains newer LEAPP packages than those shipped by https://wiki.almalinux.org/elevate/[AlmaLinux/ELevate], and support {Project} upgrades:
```
this step is not required for satellite

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+----
+# curl -o /etc/yum.repos.d/theforeman-leapp.repo https://copr.fedorainfracloud.org/coprs/g/theforeman/leapp/repo/epel-7/group_theforeman-leapp-epel-7.repo
+----
+. Install required packages:
```
for satellite, the `leapp-data-centos` package is not required

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+. We do not support EL8 installations with EPEL8 enabled, so remove `epel-release`:
++
+----
+# yum remove epel-release
```
this step is not required for satellite

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+----
++
+
+** If `preupgrade` aborts with a dependency resolution error such as:
```
this step shouldn't be required with satellite

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,159 @@
+
+[WARNING]
+====
+If you installed the system and had to use `--disable-system-checks`, the last step of the upgrade will fail, and you will need to call `foreman-installer --disable-system-checks` manually once the system is booted up.
```
this should use the `{foreman-installer}` macro

melcorr replied:
Also, we don't use language like `booted up` in official docs. 
IBM Style Guide has specifications for when and if you ever use the word "once". 
We do  not use the future tense in our docs. I've mentioned this several times by now. 
We do not use warnings in our docs unless the action with irreparably damage the deployment or cause harm to human life. This does not seem to be the case here.

bangelic replied:
Understood @melcorr T

There are other instances in the existing documentation where **NOTE** and **WARNING** have future tense in the text so I wasn't sure if that was the only time it was allowed.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+[id="Upgrading-Project-from-EL7-to-EL8{context}"]
+= Upgrading {Project} from {RHEL} 7 to {RHEL} 8
```
For Foreman, this renders as "Upgrading Foreman from Red Hat Enterprise Linux 7 to Red Hat Enterprise Linux 8", which is exactly the combination we do not support today.

Can we make it render "Upgrading Satellite from Red Hat Enterprise Linux 7 to Red Hat Enterprise Linux 8", but for the others it'd just be "Enterprise Linux" without the "Red Hat", as that would include CentOS and friends.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+LEAPP is the supported tool for {RHEL} 7 to {RHEL} 8 upgrades.
+
+.Prerequisites
+* {Project} {ProjectVersionPrevious} running on CentOS Linux 7.
```
This won't be true for Satellite, which runs on RHEL 7 only.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+. Add {Project} specific repositories to `/etc/leapp/files/leapp_upgrade_repositories.repo`:
++
+----
+# [leapp-foreman]
```
```suggestion
[leapp-foreman]
```

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+enabled=1
+gpgcheck=1
+
+[leapp-katello]
```
can we use `ifdef::katello` here inside the code block? so we hide the katello repos in the foreman docs?

allows to drop the silly "If you are not using " comment below

evgeni replied:
I used the following, and it worked:

```patch
diff --git guides/common/modules/proc_upgrading-el7-to-el8.adoc guides/common/modules/proc_upgrading-el7-to-el8.adoc
index ab89ff26..541e4aec 100644
--- guides/common/modules/proc_upgrading-el7-to-el8.adoc
+++ guides/common/modules/proc_upgrading-el7-to-el8.adoc
@@ -33,7 +33,7 @@ ifdef::katello,foreman-el[]
 . Add {Project} specific repositories to `/etc/leapp/files/leapp_upgrade_repositories.repo`:
 +
 ----
-# [leapp-foreman]
+[leapp-foreman]
 name=Foreman 3.2
 baseurl=https://yum.theforeman.org/releases/3.2/el8/$basearch
 enabled=1
@@ -64,6 +64,7 @@ gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-puppet7-release
 enabled=1
 gpgcheck=1
 
+ifdef::katello[]
 [leapp-katello]
 name=Katello 4.4
 baseurl=https://yum.theforeman.org/katello/4.4/katello/el8/$basearch/
@@ -87,11 +88,10 @@ gpgkey=https://yum.theforeman.org/pulpcore/3.16/GPG-RPM-KEY-pulpcore
 enabled=1
 gpgcheck=1
 module_hotfixes=1
+endif::[]
 ----
 endif::[]
 
-* If you are not using {Project}, you can leave out `leapp-katello`, `leapp-katello-candlepin` and `leapp-pulpcore`.
-
 * If you are using Puppet 6 instead of Puppet 7, replace the `7` with a `6` in the `leapp-puppet7` entry.
 
 * You need a Puppet repository for the Puppet agent that the installer is using.
```

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+gpgcheck=1
+module_hotfixes=1
+----
+endif::[]
```
why does this endif here? the points below still apply to katello/foreman only

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+----
+endif::[]
+
+. Remove `centos-release-scl` and `centos-release-scl-rh` repositories:
```
this should also be inside `ifdef::katello,foreman-el`

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+** If `preupgrade` aborts with a dependency resolution error such as:
++
+----
+# package rubygem-fx-0.5.0-2.el8.noarch requires rubygem(railties) >= 4.0.0,
+----
+but none of the providers can install or:
```
the errors we want to highlight are  "package rubygem-fx-0.5.0-2.el8.noarch requires rubygem(railties) >= 4.0.0, but none of the providers providers can be installed" and "package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers providers can be installed"

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+endif::[]
+
+ifdef::katello,foreman-el[]
+. Remove packages that are no longer available from any repository:
```
this needs to be a sub-item of the "dependency resolution error" step

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+
++
+
+. Check the output if it does not contain any local packages.
```
this needs to be a sub-item of the "dependency resolution error" step

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
++
+
+ifdef::katello,foreman-el[]
+** If `preupgrade` aborts with a dependency resolution error such as:
```
this whole step should look like this (step wise, not text wise):


`preupgrade` might fail with a dependency resolution error such as:

* "package rubygem-fx-0.5.0-2.el8.noarch requires rubygem(railties) >= 4.0.0, but none of the providers providers can be installed"
* "package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers providers can be installed"

if this happens, you need to clean up packages that can't be automatically upgraded, please try:
*  `dnf remove $(dnf repoquery --extras --exclude '*katello*' --exclude '*pulp*' --exclude '*localhost*' --exclude "*$HOSTNAME*" --exclude libmodulemd)`
    *  Check the output if it does not contain any local packages.
* remove `rubygem-thor`

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,184 @@
+. Remove packages that are no longer available from any repository:
++
+----
+# dnf remove $(dnf repoquery --extras --exclude '*katello*' --exclude '*pulp*' --exclude '*localhost*' --exclude "*$HOSTNAME*" --exclude libmodulemd) –
```
```suggestion
# dnf remove $(dnf repoquery --extras --exclude '*katello*' --exclude '*pulp*' --exclude '*localhost*' --exclude "*$HOSTNAME*" --exclude libmodulemd)
```

bangelic commented:
@evgeni Addressed your latest comments. Let me know if this is better.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,203 @@
+* {Project} {ProjectVersionPrevious} running on CentOS Linux 7.
+endif::[]
+
+ifdef::satellite[]
+- {Project} installations running on {RHEL} 7 cannot upgrade to {RHEL} 8. You must have a {RHEL} compatible LEAPP build to upgrade.
+endif::[]
```
This is only relevant for Foreman/Katello.
Satellite only runs on RHEL and RHEL users have the right LEAPP builds.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,203 @@
+* {Project} installations can be upgraded to CentOS Stream 8 and a {RHEL} rebuild.
+endif::[]
+
+ifdef::satellite[]
+* {Project} installations upgrade from {RHEL} 7 to {RHEL} 8, not from CentOS Linux 7 and {RHEL} rebuilds.
+endif::[]
```
Satellite is not supported on non-RHEL, so this can go away completely.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,203 @@
+gpgcheck=1
+
+
+[leapp-katello-candlepin]
+name=Candlepin: an open source entitlement management system.
+baseurl=https://yum.theforeman.org/katello/4.4/candlepin/el8/$basearch/
+gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-foreman
+enabled=1
+gpgcheck=1
+module_hotfixes=1
+
+[leapp-pulpcore]
+name=pulpcore: Fetch, Upload, Organize, and Distribute Software Packages.
+baseurl=https://yum.theforeman.org/pulpcore/3.16/el8/$basearch/
+gpgkey=https://yum.theforeman.org/pulpcore/3.16/GPG-RPM-KEY-pulpcore
+enabled=1
+gpgcheck=1
+module_hotfixes=1
```
these two should also go into the `ifdef::katello` section above.

bangelic commented:
@evgeni Addressed your latest 3 suggested changes. Is this good to be merged?

ehelms commented on guides/doc-Upgrading_and_Updating/master.adoc:
```diff
@@ -56,6 +56,11 @@ include::topics/upgrading_discovery_satellite.adoc[leveloffset=+3]
 
+ifdef::foreman-el,katello,satellite[]
+// Upgrading Foreman/Katello from EL7 to EL8
+include::common/modules/proc_upgrading-el7-to-el8.adoc[leveloffset=+2]
```
Only thought I had is that given this is a big deal to upgrade the OS and that it is required to be able to update to the following version of Foreman/katello/Satellite if it should have its own top level section header to ensure it stands out. 

Right now we have:

```
[1. Upgrade Overview]
[2. Upgrading Foreman]
[3. Updating Foreman server and Content Hosts]
```

Wondering if we need:

```
[1. Upgrade Overview]
[2. Upgrading Foreman]
[3. Upgrading Foreman from Enterprise Linux 7 to Enterprise Linux 8]
[4. Updating Foreman server and Content Hosts]
```

Lennonka replied:
@bangelic I will review after you address and resolve this comment.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,195 @@
+endif::[]
+
++
+----
```
this should be in one codeblock with the other repos in the lines above

bangelic replied:
@evgeni What specifically should be in one code block? The + sign?

evgeni replied:
All entries for the repositories should go into one

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,195 @@
+----
+
+[start=8]
+. Check the output if it does not contain any local packages.
++
+. Remove `rubygem-thor`:
++
+----
+# yum remove rubygem-thor
+----
```
this is not an own step, it still belongs to "The `preupgrade` might fail with a dependency resolution error such as"

ehelms commented:
I am totally good with doing small updates and tackling areas. Two I wanted to ask about:

 * Will we do a separate guide for Smart Proxy / Capsule upgrades?
 * Will we do a separate guide / section for disconnected to call out the differences there?

bangelic commented:
Thanks @evgeni. Pushed those changes through. Take a look and let me know if this can be merged.

evgeni commented:
I won't have time before Tuesday to further review this.

@ehelms @melcorr feel free to merge if you think it's good

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,190 @@
+[id="Upgrading-Project-from-EL7-to-EL8{context}"]
```
```suggestion
[id="Upgrading-Project-from-EL7-to-EL8_{context}"]
```

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,190 @@
+.Prerequisites
+ifdef::katello,foreman-el[]
+* {Project} {ProjectVersionPrevious} running on CentOS Linux 7.
+- {Project} installations running on {RHEL} 7 cannot upgrade to {RHEL} 8. You must have a {RHEL} compatible LEAPP build to upgrade.
```
```suggestion
* {Project} installations running on {RHEL} 7 cannot upgrade to {RHEL} 8.
You must have a {RHEL} compatible LEAPP build to upgrade.
```

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,190 @@
+.Procedure
+ifdef::katello,foreman-el[]
+. Configure the https://copr.fedorainfracloud.org/coprs/g/theforeman/leapp/[@theforeman/leapp COPR Repository], which contains newer LEAPP packages than those shipped by https://wiki.almalinux.org/elevate/[AlmaLinux/ELevate], and support {Project} upgrades:
+
```
```suggestion
```

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,190 @@
+name=Foreman 3.2
+baseurl=https://yum.theforeman.org/releases/3.2/el8/$basearch
+enabled=1
+gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-puppet7-release
```
This sounds off to me. Is is meant to be the same GPG as below?

Lennonka replied:
@bangelic Can you verify this with a SME?

evgeni replied:
I can verify that this is completely wrong and should be `gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-foreman`

bangelic replied:
@evgeni That's the verification needed haha. Thank you!

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,190 @@
+----
+# yum remove epel-release
+----
+
```
```suggestion
```

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,190 @@
+----
+# leapp preupgrade
+----
+
```
```suggestion
+
```

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,190 @@
+
+The first run is expected to fail but report issues.
+Continue to the next step for remediation.
+
+[start=7]
```
```suggestion
```

If you apply my suggestion in line 123, then defining the start should not be necessary.

maximiliankolb reviewed (commented):
Some small comments.

If there is anything mentioned in the review of your PR that you could not yet fix/add, please open an issue within github and reference the comment. Thanks!

melcorr commented:
Hey @bangelic this PR now has build errors since the last push. 
Can you please take a look and resolve any errors.

Lennonka commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,221 @@
+ifdef::satellite[]
+= Upgrading {Project} from {RHEL} 7 to {RHEL} 8
+endif::[]
+ifdef::katello,foreman-el[]
```
@maximiliankolb Would it be better to use `ifndef:satellite[]` instead?

bangelic replied:
@Lennonka I did a build preview with the `ifndef::satellite[]`and got this
![image](https://user-images.githubusercontent.com/91690569/172230172-be055f6d-e1e4-445a-acce-7b059d3d4e8a.png)
Looks like 2.5 is what would be meant for Satellite but after seeing that there is a ticket for **Upgrading Satellite from RHEL 7 to RHEL 8 with LEAPP** with accompanying documentation, can we make it so that none of this content including anything with `ifdef::{}`gets built into Satellite documentation? It just needs to be for Foreman/Katello.

maximiliankolb replied:
Yes, please apply Lena's suggestion. Something like:

```
if satellite:
  heading 1
if not satellite
  heading 2
```
Neither build targets should display empty chapters or redundant headings.

Lennonka commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,221 @@
+. Ensure `leapp preupgrade` has no issues.
+[start=2]
+. Run:
+----
```
Insert a line with `+` between the list item and the code.

Lennonka replied:
Also all other occurences.

Lennonka commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,221 @@
+# leapp answer --section remove_pam_pkcs11_module_check.confirm=True
+----
+ifdef::katello,foreman-el[]
+** The `preupgrade` might fail with a dependency resolution error such as:
+
+* "package rubygem-fx-0.5.0-2.el8.noarch requires rubygem(railties) >= 4.0.0, but none of the providers can be installed"
+* "package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers can be installed"
+
+If this happens, do the following to clean up packages that cannot automatically upgrade:
+----
+# dnf remove $(dnf repoquery --extras --exclude '*katello*' --exclude '*pulp*' --exclude '*localhost*' --exclude "*$HOSTNAME*" --exclude libmodulemd)
+----
```
```
+
The `preupgrade` might fail with a dependency resolution error such as:

* "package rubygem-fx-0.5.0-2.el8.noarch requires rubygem(railties) >= 4.0.0, but none of the providers can be installed"
* "package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers can be installed"

+
If this happens, do the following to clean up packages that cannot automatically upgrade:
+
----
# dnf remove $(dnf repoquery --extras --exclude '*katello*' --exclude '*pulp*' --exclude '*localhost*' --exclude "*$HOSTNAME*" --exclude libmodulemd)
----
```

Lennonka reviewed (changes requested):
Tips to fix the syntax

Lennonka reviewed (commented):
Okay, so the build looks good now.
I've found mentions of connected and diconnected Server -- it could use a better structure, but that's for another PR.

Lennonka commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,233 @@
+* "package rubygem-fx-0.5.0-2.el8.noarch requires rubygem(railties) >= 4.0.0, but none of the providers can be installed"
+* "package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers can be installed"
+
+If this happens, do the following to clean up packages that cannot automatically upgrade:
```
Add a line with a `+` before this so that it is inside the list item.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,233 @@
+gpgkey=https://yum.theforeman.org/pulpcore/3.16/GPG-RPM-KEY-pulpcore
+enabled=1
+gpgcheck=1
+module_hotfixes=1
```
the `ifdef::katello[]` needs to end here.

bangelic commented:
@evgeni @maximiliankolb Let me know if this can be merged. I cleaned up the conditional statements and updated the master.adoc. @Lennonka helped me with build issues that I was having and now that the conditional statements have been cleaned up, this should be good to go unless there is more content needed or needs to be updated. Thank you.

Lennonka commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,171 @@
+
+* If you are using Puppet 6 instead of Puppet 7, replace the `7` with a `6` in the `leapp-puppet7` entry.
+
+* You need a Puppet repository for the Puppet agent that the installer is using.
```
I'd add a link to the Puppet guide here.
Note that you need 2 links - sat and non-sat, and condition them.

bangelic replied:
@Lennonka Since this section is specific for Foreman/Katello and not Satellite, I will use the link for non Satellite.

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+LEAPP is the supported tool for Enterprise Linux 7 to Enterprise Linux 8 upgrades.
+
+.Prerequisites
+* {Project} {ProjectVersionPrevious} running on CentOS Linux 7.
```
Maybe it's best to say "Enterprise Linux 7" again?

If this procedure is used for both Foreman and Satellite, then maybe your can also use an attribute to differentiate between RHEL and Enterprise Linux 7. Alternatively, we can discuss setting `RHEL` to CentOS for Foreman build targets?

evgeni replied:
```suggestion
* {Project} {ProjectVersion} running on CentOS Linux 7.
```

We're upgrading the *current* version, not the previous one.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+enabled=1
+gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-foreman
+enabled=1
+gpgcheck=1
```
```suggestion
gpgcheck=1
module_hotfixes=1
```

This line was also lost somehow in all the rebases

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
++
+----
+[leapp-foreman]
+name=Foreman 3.2
+baseurl=https://yum.theforeman.org/releases/3.2/el8/$basearch
```
This document will work for Foreman 3.2 and 3.3, so I think we should use `{ProjectVersion}` here.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+
+ifdef::katello[]
+[leapp-katello]
+name=Katello 4.4
+baseurl=https://yum.theforeman.org/katello/4.4/katello/el8/$basearch/
```
This document will work for Katello 4.4 and 4.5, so I think we should use `{KatelloVersion}` here.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+
+[leapp-katello-candlepin]
+name=Candlepin: an open source entitlement management system.
+baseurl=https://yum.theforeman.org/katello/4.4/candlepin/el8/$basearch/
```
`{KatelloVersion}`

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+endif::[]
+
+[leapp-foreman-plugins]
+name=Foreman plugins 3.2
+baseurl=https://yum.theforeman.org/plugins/3.2/el8/$basearch
```
`{ProjectVersion}`

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+module_hotfixes=1
+
+[leapp-foreman-client]
+name=Foreman client 3.2
+baseurl=https://yum.theforeman.org/client/3.2/el8/$basearch
```
`{ProjectVersion}`

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+
+[leapp-pulpcore]
+name=pulpcore: Fetch, Upload, Organize, and Distribute Software Packages.
+baseurl=https://yum.theforeman.org/pulpcore/3.16/el8/$basearch/
+gpgkey=https://yum.theforeman.org/pulpcore/3.16/GPG-RPM-KEY-pulpcore
```
we don't have a macro for the pulpcore version (the user usually doesn't have to set it themself).
Katello 4.4 is using 3.16, 4.5 is using 3.17 (well, soon 3.18 I guess) -- should we introduce a macro or have an if/else here directly?

evgeni replied:
This is still outstanding, I think adding a new macro is the simplest?

evgeni replied:
```diff
diff --git guides/common/attributes.adoc guides/common/attributes.adoc
index 65eec3bb..f2413a79 100644
--- guides/common/attributes.adoc
+++ guides/common/attributes.adoc
@@ -6,6 +6,7 @@
 :ProjectVersion: nightly
 :ProjectVersionPrevious: 3.3
 :KatelloVersion: nightly
+:PulpcoreVersion: 3.18
 :TargetVersion: {ProjectVersion}
 :TargetVersionMaintainUpgrade: {ProjectVersion}
 // The above attribute should point to the GA version number (x.y) for all releases including Beta
diff --git guides/common/modules/proc_upgrading-el7-to-el8.adoc guides/common/modules/proc_upgrading-el7-to-el8.adoc
index 1940d7d2..45988e87 100644
--- guides/common/modules/proc_upgrading-el7-to-el8.adoc
+++ guides/common/modules/proc_upgrading-el7-to-el8.adoc
@@ -61,8 +61,8 @@ module_hotfixes=1
 
 [leapp-pulpcore]
 name=pulpcore: Fetch, Upload, Organize, and Distribute Software Packages.
-baseurl=https://yum.theforeman.org/pulpcore/3.16/el8/$basearch/
-gpgkey=https://yum.theforeman.org/pulpcore/3.16/GPG-RPM-KEY-pulpcore
+baseurl=https://yum.theforeman.org/pulpcore/{PulpcoreVersion}/el8/$basearch/
+gpgkey=https://yum.theforeman.org/pulpcore/{PulpcoreVersion}/GPG-RPM-KEY-pulpcore
 enabled=1
 gpgcheck=1
```

something like this should do (we then need to adjust the attribute differently in the branches, but that's fine, that's what attributes are for)

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+
+* You need a Puppet repository for the Puppet agent that the installer is using.
+
+For more information, see link:https://docs.theforeman.org/nightly/Managing_Configurations_Puppet/index-katello.html[Configuring Hosts Using Puppet].
```
I don't see how this link is relevant to the upgrade?

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+[leapp-foreman]
+name=Foreman 3.2
+baseurl=https://yum.theforeman.org/releases/3.2/el8/$basearch
+enabled=1
```
```suggestion
```

this line is duplicated, accidentally

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+* "package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers can be installed"
+
+If this happens, do the following to clean up packages that cannot automatically upgrade:
++
```
these `+` render weirdly for me:

![image](https://user-images.githubusercontent.com/94284/172556400-982b5ee6-4686-4b00-9b75-95144796cbc3.png)

bangelic replied:
Same. I will delete the +.

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+# yum install leapp leapp-repository leapp-data-centos
+----
+
+[start=3]
```
I am still not convinced we really need this. It looks very error-prone in the long run. Usually, when you exclude certain steps for certain built targets, the lists in asciidoc still work as expected.

maximiliankolb commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+
+[start=10]
+. Reboot the system.
+
+[start=11]
```
```suggestion
```

You don't need any empty lines between two list elements.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+* "package rubygem-fx-0.5.0-2.el8.noarch requires rubygem(railties) >= 4.0.0, but none of the providers can be installed"
+* "package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers can be installed"
+
+If this happens, do the following to clean up packages that cannot automatically upgrade:
++
+----
+# dnf remove $(dnf repoquery --extras --exclude '*katello*' --exclude '*pulp*' --exclude '*localhost*' --exclude "*$HOSTNAME*" --exclude libmodulemd)
+----
+
+Check the output if it does not contain any local packages.
+
+Remove `rubygem-thor`:
++
+----
+# yum remove rubygem-thor
+----
```
After a bit of thinking, the `dnf remove $(dnf repoquery …)` step is too dangerous. Let's make 

```suggestion
If this happens, do the following to clean up packages that cannot automatically upgrade (`rubygem(thor)` and `rubygem(railties)` in the example above):
+
----
# yum remove rubygem-thor rubygem-railties
----
```

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+* {Project} installations can be upgraded to CentOS Stream 8 and a {RHEL} rebuild.
+
+.Procedure
+[start=1]
```
is there a way to avoid those `start=X` lines? it's horrible to read and will make us miss updating them if we ever change the procedure

bangelic replied:
@evgeni I had to insert the `start=x` lines because the steps kept restarting at 1 throughout the document due to code blocks and other sentences interrupting the flow of the procedure.

maximiliankolb replied:
I super rarely use this. IMHO it's not needed here. Please remove all empty lines between list elements and remove all instances of "start=X".

bangelic replied:
Can we compromise? I will only use `start=x` starting at the 8th step because I have tried playing around with the + sign and attempted dozens of build previews and it will not continue the numbering after the 7th step. The + signs end up in the text when I do a build preview and it doesn't matter where I put them. The numbering starts over at 1 after number 7.

bangelic replied:
Another step was recommended and added. Now the `start=x` will begin at the 9th step if you guys are cool with it.

bangelic replied:
Never mind. Vendula helped me fix the issue I was having. I will take out all `start=x` references.

evgeni reviewed (commented):
Let's support RHEL upgrades too.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,173 @@
+
+.Prerequisites
+* {Project} {ProjectVersionPrevious} running on CentOS Linux 7.
+* {Project} installations running on {RHEL} 7 cannot upgrade to {RHEL} 8.
+You must have a {RHEL} compatible LEAPP build to upgrade.
```
Why was this marked as resolved?

bangelic replied:
Shouldn't have been. It's resolved for real now.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,170 @@
+----
+# yum remove rubygem-thor rubygem-railties
+----
++
+Check the output if it does not contain any local packages.
++
+Remove `rubygem-thor`:
++
+----
+# yum remove rubygem-thor
+----
```
```suggestion
```

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,170 @@
+----
+. Reboot the system.
+
+. After the system reboots, a live system conducts the upgrade, reboots to fix SELinux labels, then reboots into the final Enterprise Linux 8 system.
+. After the system reboots, LEAPP finishes the upgrade, watch it with:
```
starting two points with "After the system reboots" sounds odd to me, but I am not sure how one could re-word this better.

bangelic replied:
@evgeni I made it easy. Just removed "After the system reboots" and it still makes sense.

bangelic commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,170 @@
++
+The `preupgrade` might fail with a dependency resolution error such as:
++
+----
+* "package rubygem-fx-0.5.0-2.el8.noarch requires rubygem(railties) >= 4.0.0, but none of the providers can be installed"
+* "package rubygem-railties-6.0.4.7-1.el8.noarch requires rubygem(thor) < 2.0, but none of the providers can be installed"
+----
```
Fixed

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,162 @@
+----
+
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+. Add {Project} specific repositories to `/etc/leapp/files/leapp_upgrade_repositories.repo`:
++
```
```suggestion
. Add {Project} specific repositories to `/etc/leapp/files/leapp_upgrade_repositories.repo`:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
```

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,162 @@
+----
+[leapp-foreman]
+name=Foreman {ProjectVersion}
+baseurl=https://yum.theforeman.org/releases/{ProjectVersion}/3.2/el8/$basearch
```
```suggestion
baseurl=https://yum.theforeman.org/releases/{ProjectVersion}/el8/$basearch
```

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,162 @@
+The first run is expected to fail but report issues.
+Continue to the next step for remediation.
+
+. Fix issues found by LEAPP (check `/var/log/leapp/leapp-report.txt` for details):
```
```suggestion
. Examine the report in the `/var/log/leapp/leapp-report.txt` file, answer all questions (using `leapp answer`) and manually resolve the other reported problems. The following commands show the most common steps required:
```

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,162 @@
+* {Project} {ProjectVersion} running on Enterprise Linux 7.
+* Access to available repositories or a local mirror of repositories.
+* No outstanding updates.
+* {Project} installations can be upgraded to CentOS Stream 8 and a {RHEL} rebuild.
```
```suggestion
* {Project} installations running on CentOS 7 can be upgraded to CentOS Stream 8 or a {RHEL} rebuild.
* {Project} installations running on {RHEL} 7 can be upgraded to {RHEL} 8.
```

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,162 @@
+* {Project} installations can be upgraded to CentOS Stream 8 and a {RHEL} rebuild.
+
+.Procedure
+. Do not perform this step on {RHEL} as this is for CentOS. Configure the https://copr.fedorainfracloud.org/coprs/g/theforeman/leapp/[@theforeman/leapp COPR Repository], which contains newer LEAPP packages than those shipped by https://wiki.almalinux.org/elevate/[AlmaLinux/ELevate], and support {Project} upgrades:
++
+----
+# curl -o /etc/yum.repos.d/theforeman-leapp.repo https://copr.fedorainfracloud.org/coprs/g/theforeman/leapp/repo/epel-7/group_theforeman-leapp-epel-7.repo
+----
```
```suggestion
. Configure the repositories to obtain LEAPP.
+
On CentOS, configure the https://copr.fedorainfracloud.org/coprs/g/theforeman/leapp/[@theforeman/leapp COPR Repository], which contains newer LEAPP packages than those shipped by https://wiki.almalinux.org/elevate/[AlmaLinux/ELevate], and support {Project} upgrades:
+
----
# curl -o /etc/yum.repos.d/theforeman-leapp.repo https://copr.fedorainfracloud.org/coprs/g/theforeman/leapp/repo/epel-7/group_theforeman-leapp-epel-7.repo
----
+
On {RHEL}, enable the `rhel-7-server-extras-rpms` repository:
+
----
# subscription-manager repos --enable=rhel-7-server-extras-rpms
----
```

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,162 @@
+# yum install leapp leapp-repository
+----
+
+. Install additional OS specific packages (`leapp-data-centos` for CentOS Stream, `leapp-data-rocky` for Rocky Linux, `leapp-data-almalinux` for AlmaLinux`). Not required for {RHEL}.
```
```suggestion
. Install additional OS specific packages (`leapp-data-centos` for CentOS Stream, `leapp-data-rocky` for Rocky Linux, `leapp-data-almalinux` for AlmaLinux). Not required for {RHEL}.
```

bangelic commented:
"This is still outstanding, I think adding a new macro is the simplest?"
@evgeni Do you have a macro you want to use?

evgeni commented:
> "This is still outstanding, I think adding a new macro is the simplest?"
> @evgeni Do you have a macro you want to use?

I could swear I posted a follow up comment with an example patch. Can't find it. Stupid GitHub UI folding comments left and right.

I have the following patch applied locally:

```diff
--- guides/common/attributes.adoc
+++ guides/common/attributes.adoc
@@ -6,6 +6,7 @@
 :ProjectVersion: nightly
 :ProjectVersionPrevious: 3.3
 :KatelloVersion: nightly
+:PulpcoreVersion: 3.18
 :TargetVersion: {ProjectVersion}
 :TargetVersionMaintainUpgrade: {ProjectVersion}
 // The above attribute should point to the GA version number (x.y) for all releases including Beta
diff --git guides/common/modules/proc_upgrading-el7-to-el8.adoc guides/common/modules/proc_upgrading-el7-to-el8.adoc
index 1940d7d2..45988e87 100644
--- guides/common/modules/proc_upgrading-el7-to-el8.adoc
+++ guides/common/modules/proc_upgrading-el7-to-el8.adoc
@@ -61,8 +61,8 @@ module_hotfixes=1
 
 [leapp-pulpcore]
 name=pulpcore: Fetch, Upload, Organize, and Distribute Software Packages.
-baseurl=https://yum.theforeman.org/pulpcore/3.16/el8/$basearch/
-gpgkey=https://yum.theforeman.org/pulpcore/3.16/GPG-RPM-KEY-pulpcore
+baseurl=https://yum.theforeman.org/pulpcore/{PulpcoreVersion}/el8/$basearch/
+gpgkey=https://yum.theforeman.org/pulpcore/{PulpcoreVersion}/GPG-RPM-KEY-pulpcore
 enabled=1
 gpgcheck=1
 module_hotfixes=1
```

Read: add `:PulpcoreVersion: 3.18` to `attributes.adoc` and then use `{PulpcoreVersion}` in the code block.

evgeni commented on guides/common/modules/proc_upgrading-el7-to-el8.adoc:
```diff
@@ -0,0 +1,171 @@
+* {Project} {ProjectVersion} running on Enterprise Linux 7.
+* Access to available repositories or a local mirror of repositories.
+* No outstanding updates.
+* {Project} installations running on CentOS 7 can be upgraded to CentOS Stream 8 or a {RHEL} rebuild.
```
```suggestion
* {Project} installations running on CentOS 7 can be upgraded to CentOS Stream 8 or a {RHEL} rebuild.
* {Project} installations running on {RHEL} 7 can be upgraded to {RHEL} 8.
```

evgeni reviewed (approved):
There are two outstanding things, which I do not consider blocking:
* the link to the Puppet guide doesn't seem to add value to the uprgrade process? https://github.com/theforeman/foreman-documentation/pull/1248/files#r892002478
* we should mention that RHEL7→RHEL8 does work now. https://github.com/theforeman/foreman-documentation/pull/1248/files#r894551939

@Lennonka did you want to do any further review on this one?

evgeni commented:
I am hearing Lena has no time today, so I'll just merge this, and then we see from there.

maximiliankolb commented:
Is there anything left to do here? Do we need to create issue? @bangelic 

In general, I feel like this PR needs some more love.
For example, the title does not match the anchor and file name; and some lines don't really match our style, like `* No outstanding updates.`.

cc @melcorr
